relay-cardinality/src/redis.rs

```diff
@@ -0,0 +1,429 @@
+struct Scope {
+    /// The organization.
+    pub organization_id: OrganizationId,
+    /// The metric namespace, extracted from the metric bucket name.
+    pub namespace: MetricNamespace,
+}
+
+/// Implementation uses Redis sets to keep track of cardinality.
+pub struct RedisSetLimiter {
+    redis: RedisPool,
+    window: SlidingWindow,
+    #[cfg(test)]
+    time_offset: u64,
+}
+
+/// A Redis based limiter using Redis sets to track cardinality and membership.
+impl RedisSetLimiter {
+    /// Creates a new [`RedisSetLimiter`].
+    pub fn new(redis: RedisPool, window: SlidingWindow) -> Self {
+        Self {
+            redis,
+            window,
+            #[cfg(test)]
+            time_offset: 0,
+        }
+    }
+}
+
+impl RedisSetLimiter {
+    /// Checks the limits for a specific scope.
+    ///
+    /// Returns an iterator over all entries which have been accepted.
+    fn check_limits(
+        &self,
+        client: &mut RedisHelper,
+        scope: Scope,
+        mut entries: Vec<Entry>,
+        timestamp: u64,
+        limit: usize,
+    ) -> Result<impl Iterator<Item = Entry>> {
+        metric!(
+            counter(CardinalityLimiterCounters::RedisRead) += 1,
+            namespace = scope.namespace.as_str()
+        );
+        metric!(
+            counter(CardinalityLimiterCounters::RedisHashCheck) += entries.len() as i64,
+            namespace = scope.namespace.as_str()
+        );
+
+        let active_window = client
+            .read_set(&self.to_redis_key(&scope, self.window.active_time_bucket(timestamp)))?;
+        let mut new_hashes = HashSet::new();
+
+        let mut new_cardinality = active_window.len();
+        let mut rejected = 0;
+
+        for entry in &mut entries {
+            if active_window.contains(&entry.hash) {
+                entry.accept();
+            } else if new_cardinality < limit {
+                entry.accept();
+                new_hashes.insert(entry.hash);
+                new_cardinality += 1;
+            } else {
+                entry.reject();
+                rejected += 1;
+            }
+        }
+
+        if rejected > 0 {
+            relay_log::debug!(
+                organization_id = scope.organization_id,
+                namespace = scope.namespace.as_str(),
+                "rejected {rejected} metrics due to cardinality limit"
+            );
+            metric!(
+                counter(CardinalityLimiterCounters::Rejected) += rejected,
+                namespace = scope.namespace.as_str()
+            );
+        }
+
+        if !new_hashes.is_empty() {
+            for time_bucket in self.window.iter(timestamp) {
+                let key = self.to_redis_key(&scope, time_bucket);
+
+                // The expiry is a off by `window.granularity_seconds`,
+                // but since this is only used for cleanup, this is not an issue.
+                client.add_to_set(&key, &new_hashes, self.window.window_seconds)?;
```

Comment:
We could start the redis pipeline outside of this loop and send all commands at once.

Reply:
Can the pipeline become too big?

Reply:
Good question, did not consider that.